### PR TITLE
Import SonarQube security hotspots

### PR DESCRIPTION
**Pull request by jimtsikos**
This PR imports security hotspots using the SonarQube API.

### CODE REVIEW

**Comment by valentijnscholten** [comment]
Thanks for the PR, but I have some questions:
- in #4068 we added the possibility to import all types of findings/issues from sonarqube. is there missing some that warrants a new PR?
- I see a lot of copy and pasting of code from the existing sonarqube importer, is there anyway to avoid that. For example by paramatrizing a method that does the actual import?

**Comment by jimtsikos** [comment]
@valentijnscholten I would be glad to answer your questions. 

- Unfortunately, the API call **api/issues/search** does not return the security hotspots. The only acceptable types are: bugs, code smells and vulnerabilities. To be able and get the security hotspots I used the internal API call **api/hotspots/search**. In addition, in order to get the hotspot rule I used the API call **api/hotspots/show**. 
- I understand your concerns about the code. I chose to create different methods for hotspots and not a common parameterized method because hotspot is a different entity and may require changes in the future that issues will not need. If what I thought is not acceptable, then I could refactor the code (any suggestions are more than welcome) with a parameterized method, it is not a big deal. Just let me know what you think about it.

**Comment by damiencarol** [comment]
@jimtsikos do you know if an official python module to use SonarQube API exists? I have the feeling that we could not be the only only ones to use SonarQube API in a Python programme.

**Comment by jimtsikos** [comment]
@damiencarol the truth is that I did not find an official python module for SonarQube API, no matter how much I searched.

**Comment by damiencarol** [comment]
ouch. ok... :D

**Review by valentijnscholten — APPROVED** [review]
migration needs an update

**Comment by jimtsikos** [comment]
@valentijnscholten db migration updated

**Comment by jimtsikos** [comment]
Hi @valentijnscholten, do you have any updates on this PR review? Should I contact another reviewer?

File: dojo/tools/sonarqube_api/importer.py
Comment (damiencarol) [inline dojo/tools/sonarqube_api/importer.py:207]: You don't need to set every attribute.
```suggestion
```
`mitigation`, `impact` are not mandatory.
Also `numerical_severity` is calculated in upper layers.

**Comment by jimtsikos** [comment]
> @jimtsikos we already merged few DB migration scripts. You need to rename the source of your DB migration part.

@damiencarol Both requested changes done.

**Comment by damiencarol** [comment]
@jimtsikos any news about this PR? Are you still working on it?

**Comment by tarrinho** [comment]
Any news?

**Comment by jimtsikos** [comment]
@damiencarol and @tarrinho I will be working on the requested changes this week.

**Comment by Maffooch** [comment]
Hi @jimtsikos can you please [allow edits from maintainers](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork) so we can update conflicts for you and get this merged quicker? There is a lot of interest in your contribution :)

**Comment by jimtsikos** [comment]
Hi, @Maffooch thank you very much for your offer. Though, I do not have the necessary permissions to allow edits from maintainers. Do you know who may have the required permissions and allow the edits? In the meantime, I will try to resolve the conflicts.

**Comment by Maffooch** [comment]
@jimtsikos Oh I just noticed that the fork this PR is based from is not your fork. I would think you would need write permissions to that fork to grant permission to maintainers to edit? Not totally sure.

**Comment by nu-vi** [comment]
Any updates on this?

**Comment by Maffooch** [comment]
@jimtsikos Please see conflicts and update migrations

File: dojo/tools/sonarqube_api/importer.py
Comment (jimtsikos) [reply dojo/tools/sonarqube_api/importer.py]: @damiencarol I have made the changes requested using the textwrap library

**Comment by jimtsikos** [comment]
Hi guys, sorry for the delay, I finally made the changes. To be more specific:
@damiencarol I made the changes you asked for (textwrap). So, you can review the code.
@Maffooch I believe that I merged the conflicts successfully and I have already updated my code and unit tests. Can you please verify that the merging of conflicts was successful as well?

File: dojo/unittests/tools/test_sonarqube_importer.py
Comment (damiencarol) [inline dojo/unittests/tools/test_sonarqube_importer.py:351]: Please add more checks on the data.

File: dojo/tools/sonarqube_api/importer.py
Comment (damiencarol) [inline dojo/tools/sonarqube_api/importer.py:23]: Hotspots generate a lot of false positive.
Could you add a feature flag that enable/disable this line?
```suggestion
        if settings.SONARQUBE_API_PARSER_HOTSPOTS:
            items.extend(self.import_hotspots(test))
```

**Review by damiencarol — CHANGES_REQUESTED** [review]

File: dojo/tools/sonarqube_api/importer.py
Comment (damiencarol) [reply dojo/tools/sonarqube_api/importer.py:23]: You have an example here:
https://github.com/DefectDojo/django-DefectDojo/blob/92a14eac19ab6c53777c802f334d1b9c48a5b54a/dojo/settings/settings.dist.py#L165-L167

File: dojo/unittests/tools/test_sonarqube_importer.py
Comment (jimtsikos) [reply dojo/unittests/tools/test_sonarqube_importer.py:351]: @damiencarol can you please be more specific on the checks required? An example or two would be really helpful!

Comment (Maffooch) [reply dojo/unittests/tools/test_sonarqube_importer.py:351]: Some other things to check would be title, severity, CWE/CVE, etc.
Just other things that are fields in the finding object. This is to help maintain the parser to little changes that add up over time, or possibly a change to the report/api format.

Comment (jimtsikos) [reply dojo/unittests/tools/test_sonarqube_importer.py:495]: Ok, I see, thanks for the clarification.

Comment (jimtsikos) [reply dojo/unittests/tools/test_sonarqube_importer.py:495]: Done

File: dojo/tools/sonarqube_api/importer.py
Comment (jimtsikos) [reply dojo/tools/sonarqube_api/importer.py]: Done

**Comment by jimtsikos** [comment]
@damiencarol and @Maffooch I completed the changes requested. Please proceed with the review.

**Review by Maffooch — APPROVED** [review]

File: dojo/tools/sonarqube_api/importer.py
Comment (damiencarol) [inline dojo/tools/sonarqube_api/importer.py:231]: 1. Don't send notification.
2. The exception is swallowed. It will return empty list that will close existing findings with some use cases.
You can remove the try/catch

File: dojo/unittests/tools/test_sonarqube_importer.py
Comment (damiencarol) [inline dojo/unittests/tools/test_sonarqube_importer.py:92]: Something is wrong here.
We are testing exactly how the code of the parser analyze a dict.
It's too low level.
You need to specialize one unit test with the data of a specific report.

File: dojo/tools/sonarqube_api/importer.py
Comment (damiencarol) [inline dojo/tools/sonarqube_api/importer.py:187]: This is not the good value for severity/
According to the SQ documentation:

> Security Hotspots are not assigned severities as it is unknown whether there is truly an underlying vulnerability until they are reviewed.

Security Hotspots are infos to investigate. The data in `vulnerabilityProbability` is the equivalent of the attribute `scanner_confidence` in the current implementation.

File: dojo/tools/sonarqube_api/importer.py
Comment (damiencarol) [inline dojo/tools/sonarqube_api/importer.py:182]: You don't need this complex if/else.
`textwrap.shorten()` have this logic inside.
```suggestion
                title = textwrap.shorten(text=hotspot['message'], width=500)
```

**Review by damiencarol — CHANGES_REQUESTED** [review]

File: dojo/tools/sonarqube_api/importer.py
Comment (damiencarol) [reply dojo/tools/sonarqube_api/importer.py:182]:
```bash
>>> shorten("1 3 5 7 9", 8)
'1 [...]'
>>> len(shorten("1 3 5 7 9", 8))
7
```

File: dojo/tools/sonarqube_api/importer.py
Comment (jimtsikos) [reply dojo/tools/sonarqube_api/importer.py:187]: So, what do you propose to do? Should I leave the severity empty or hardcoded as Info?

File: dojo/unittests/tools/test_sonarqube_importer.py
Comment (jimtsikos) [reply dojo/unittests/tools/test_sonarqube_importer.py:92]: @damiencarol So, just to clarify the comment above:

1. Should I leave the tests to check the number of findings and create a new test to check the report fields?
2. Let's take for example the title in a finding, should I test it against the string itself and not from the loaded issue or hotspot?

File: dojo/tools/sonarqube_api/importer.py
Comment (damiencarol) [reply dojo/tools/sonarqube_api/importer.py:187]: Simply `Info` because for SonarQube it is info. But you can use this field to populate the confidence attribute.

File: dojo/unittests/tools/test_sonarqube_importer.py
Comment (damiencarol) [reply dojo/unittests/tools/test_sonarqube_importer.py:92]: You can just test the fields of the first findings that come from hotpots data in the test below but consider the `get_findings` method as a black-box.
If you make assumptions on which field is read and link it in the unit test your are just re-writing the code of the parser in the unit tests. Which means it's duplicate code, which means 2x the work if the data structure of SQ is changing or if we modify the implementation of the parser.
Just add few more checks on the result of `get_findings` method in the news tests you wrote.

File: dojo/tools/sonarqube_api/importer.py
Comment (jimtsikos) [reply dojo/tools/sonarqube_api/importer.py]: Done

File: dojo/tools/sonarqube_api/importer.py
Comment (jimtsikos) [reply dojo/tools/sonarqube_api/importer.py]: Done

File: dojo/tools/sonarqube_api/importer.py
Comment (jimtsikos) [reply dojo/tools/sonarqube_api/importer.py]: Done

**Comment by jimtsikos** [comment]
@damiencarol Can you please help me to spot why the unit tests are failing? By viewing the logs I imagine that it has something to do with Jira, but I am not sure.

**Comment by damiencarol** [comment]
Will take a look today. Is it ok if I push fix directly in the branch?

**Comment by jimtsikos** [comment]
Yes, it is ok to push fix directly in the branch.

File: dojo/unittests/tools/test_sonarqube_importer.py
Comment (jimtsikos) [reply dojo/unittests/tools/test_sonarqube_importer.py]: Done, if I understood it correctly.

**Comment by damiencarol** [comment]
I'm able to reproduce the bug locally

**Comment by damiencarol** [comment]
@jimtsikos I understand why it's blocked. You need to mock also in an existing test all the new call to the hotspots endpoints. Easy fix

**Comment by damiencarol** [comment]
@jimtsikos You created the pull request from your dev branch and not a feature branch. So I can't push to it :D 
Would you mind applying this patch to your branch 

[test_sonarqube_api_parser.py.patch](https://github.com/DefectDojo/django-DefectDojo/files/7227019/test_sonarqube_api_parser.py.patch.txt)
?

File: dojo/unittests/tools/test_sonarqube_api_parser.py
Comment (damiencarol) [inline dojo/unittests/tools/test_sonarqube_api_parser.py:31]: @jimtsikos missed one blank line.

Comment (jimtsikos) [reply dojo/unittests/tools/test_sonarqube_api_parser.py:31]: Done

**Comment by jimtsikos** [comment]
@damiencarol You can review the changes

**Review by damiencarol — APPROVED** [review]

File: dojo/db_migrations/0123_sonarqube_api_type_length_change.py
Comment (StefanFl) [inline dojo/db_migrations/0123_sonarqube_api_type_length_change.py:9]:
```suggestion
        ('dojo', '0123_scan_type'),
```

There has been another migration merged in dev

**Review by StefanFl — CHANGES_REQUESTED** [review]

File: dojo/db_migrations/0123_sonarqube_api_type_length_change.py
Comment (StefanFl) [reply dojo/db_migrations/0123_sonarqube_api_type_length_change.py:9]: @jimtsikos Will merge it directly when this is resolved 😃

Comment (jimtsikos) [reply dojo/db_migrations/0123_sonarqube_api_type_length_change.py]: @StefanFl Changes applied.

**Comment by damiencarol** [comment]
pushed the change to move forward with this PR

**Pull request merged**
